frameworks/file-analysis.rst

```diff
@@ -126,38 +161,77 @@ calculate the MD5 of plain text files:
+separate file on disk. The output file name results from concatenating the
+:zeek:see:`FileExtract::prefix` (normally ``./extract_files/``) and the
+enumerated ``file-NNNN`` strings.
+
+In a production setting you'll likely want to include additional information in
+the output, for example from state attached to the provided file record. The
+Zeek distribution ships with a starting point for such approaches: the
+``policy/frameworks/files/extract-all-files.zeek`` script. For additional
+configurability, take a look at the `file-extraction
+<https://github.com/hosom/file-extraction>`_ Zeek package.
+
+Content analysis
+^^^^^^^^^^^^^^^^
+
+The ``FileDataEvent`` analyzer provides script-layer access to file content for
+customized analysis. Since observed files can be very large, Zeek cannot buffer
+these files and provide their entire content to the script layer once
+complete. Instead, the ``FileDataEvent`` analyzer reflects the incremental
+nature of file content as Zeek observes it, and supports two types of events to
+allow you to process it: user-provided `stream events` receive new file content
+as supplied by connection-oriented protocols, while `chunk events` receive
+observed data as provided by protocols that do not feature stream semantics.
+
+The following example manually computes the SHA256 hash of each observed file by
+building up hash state and feeding streamed file content into the hash
+computation. When Zeek removes a file's state (because it has fully observed it,
+or perhaps because its state is timing out), it prints the resulting hash to the
+console:
 
-You may add the same analyzer type multiple times to a given file, assuming you
-use varying :zeek:see:`Files::AnalyzerArgs` parameterization, and remove them
-selectively from files via calls to :zeek:see:`Files::remove_analyzer`. You may
-also enable and disable file analyzers globally by calling
-:zeek:see:`Files::enable_analyzer` and :zeek:see:`Files::disable_analyzer`,
-respectively.
+.. code-block:: zeek
 
-For additional customizations and APIs, please refer to
-:doc:`base/frameworks/files/main.zeek </scripts/base/frameworks/files/main.zeek>`.
+    global hashstate: table[string] of opaque of sha256;
 
-Regardless of which file analyzers end up acting on a file, general
-information about the file (e.g. size, time of last data transferred,
-MIME type, etc.) is logged in ``files.log``.
+    event file_stream(f: fa_file, data: string)
+        {
+        if ( f$id !in hashstate )
+            hashstate[f$id] = sha256_hash_init();
+
+        sha256_hash_update(hashstate[f$id], data);
+        }
+
+    event file_new(f: fa_file)
+        {
+        Files::add_analyzer(f, Files::ANALYZER_DATA_EVENT, [$stream_event=file_stream]);
+        }
+
+    event file_state_remove(f: fa_file)
+        {
+        if ( f$id in hashstate )
+            {
+            print(sha256_hash_finish(hashstate[f$id]));
+            delete hashstate[f$id];
+            }
+        }
```

Comment:
There have been some perf issues with ANALZYER_DATA_EVENT in production systems where the analyzer wasn't registered. Think multiple GB files triggering lots of events (and maybe only the first few kb are actually of interest).

It is also fairly dangerous to start buffering the data in script-land just because one can.

Not sure these need to be pointed out, but leaving the comment anyway.

Reply:
Ah, right — thank you for flagging that. I'll add that.

I'll also add a bit about the fact that protocol analyzers can redef additional state onto the file record, like in `base/protocols/http/entities.zeek`.